osquery/tables/system/linux/processes.cpp

```diff
@@ -55,12 +60,46 @@ inline std::string readProcLink(const std::string& attr,
   char link_path[PATH_MAX] = {0};
   auto bytes = readlink(attr_path.c_str(), link_path, sizeof(link_path) - 1);
   if (bytes >= 0) {
+    link_path[bytes] = 0; // readlink does not append NUL
     result = std::string(link_path);
   }
 
   return result;
 }
 
+// In the case where the linked binary path ends in " (deleted", and a file
+// actually exists at that path, check whether the inode of that file matches
+// the inode of the mapped file in /proc/%pid/maps
+Status deletedMatchesInode(const std::string& path, const std::string& pid) {
+  const std::string maps_path = getProcAttr("maps", pid);
+  std::string maps_contents;
+  auto s = osquery::readFile(maps_path, maps_contents);
+  if (!s.ok()) {
+    return Status(-1, "Cannot read maps file: " + maps_path);
+  }
+
+  // Extract the expected inode of the binary file from /proc/%pid/maps
+  boost::smatch what;
+  boost::regex expression("([0-9]+)\\h+\\Q" + path + "\\E");
+  if (!boost::regex_search(maps_contents, what, expression)) {
+    return Status(-1, "Could not find binary inode in maps file: " + maps_path);
+  }
+  std::string inode = what[1];
+
+  // stat the file at the expected binary path
+  struct stat st;
+  if (stat(path.c_str(), &st) != 0) {
+    return Status(-1, "Error in stat of binary: " + path);
+  }
+
+  // If the inodes match, the binary name actually ends with " (deleted)"
+  if (boost::lexical_cast<std::string>(st.st_ino) == inode) {
```

Comment:
We ran some performance tests a while back and the `std::to_string` casting is much faster :)